jupyter_server/services/contents/filemanager.py

```diff
@@ -55,6 +56,31 @@ def _validate_root_dir(self, proposal):
             raise TraitError("%r is not a directory" % value)
         return value
 
+    @default("preferred_dir")
+    def _default_preferred_dir(self):
+        try:
+            value = self.parent.preferred_dir
+        except AttributeError:
+            pass
+        else:
+            if value is not None:
+                warnings.warn(
+                    "ServerApp.preferred_dir config is deprecated in jupyter-server 2.0. Use ContentsManager.preferred_dir with a relative path instead",
+                    DeprecationWarning,
+                    stacklevel=3,
+                )
+                if not (value + os.path.sep).startswith(self.root_dir):
+                    raise TraitError("%s is outside root contents directory" % value)
+                return os.path.relpath(value, self.root_dir).replace(os.path.sep, "/")
+        return "/"
+
+    @validate("preferred_dir")
+    def _validate_preferred_dir(self, proposal):
+        try:
+            return super()._validate_preferred_dir(proposal)
+        except HTTPError as e:
+            raise TraitError(e.log_message) from e
+
```

Comment:
This block isn't _currently_ necessary. 
```suggestion
```

Reply:
While improved error handling isn't strictly *needed*, it can still be *helpful*. E.g. if I pass the argument `--ContentsManager.preferred_dir=../foo`, then with the above code I get:

```
[C 2022-10-15 14:07:03.177 ServerApp] Bad config encountered during initialization: ../example is outside root contents directory
```

Removing it, I get a much harder to parse:

```
Traceback (most recent call last):
  File "<env dir>\Scripts\jupyter-lab-script.py", line 9, in <module>
    sys.exit(main())
  File "<dev dir>\jupyter_server\jupyter_server\extension\application.py", line 598, in launch_instance
    serverapp = cls.initialize_server(argv=args)
  File "<dev dir>\jupyter_server\jupyter_server\extension\application.py", line 568, in initialize_server
    serverapp.initialize(
  File "<env dir>\lib\site-packages\traitlets\config\application.py", line 110, in inner
    return method(app, *args, **kwargs)
  File "<dev dir>\jupyter_server\jupyter_server\serverapp.py", line 2477, in initialize
    self.init_configurables()
  File "<dev dir>\jupyter_server\jupyter_server\serverapp.py", line 1834, in init_configurables
    self.contents_manager = self.contents_manager_class(
  File "<env dir>\lib\site-packages\traitlets\config\configurable.py", line 109, in __init__
    self.config = config
  File "<env dir>\lib\site-packages\traitlets\traitlets.py", line 715, in __set__
    self.set(obj, value)
  File "<env dir>\lib\site-packages\traitlets\traitlets.py", line 704, in set
    obj._notify_trait(self.name, old_value, new_value)
  File "<env dir>\lib\site-packages\traitlets\traitlets.py", line 1374, in _notify_trait
    self.notify_change(
  File "<env dir>\lib\site-packages\traitlets\traitlets.py", line 1386, in notify_change
    return self._notify_observers(change)
  File "<env dir>\lib\site-packages\traitlets\traitlets.py", line 1431, in _notify_observers
    c(event)
  File "<env dir>\lib\site-packages\traitlets\traitlets.py", line 1045, in compatible_observer
    return func(self, change)
  File "<env dir>\lib\site-packages\traitlets\config\configurable.py", line 218, in _config_changed
    self._load_config(change.new, traits=traits, section_names=section_names)
  File "<env dir>\lib\site-packages\traitlets\config\configurable.py", line 200, in _load_config
    warn(msg)
  File "<env dir>\lib\contextlib.py", line 124, in __exit__
    next(self.gen)
  File "<env dir>\lib\site-packages\traitlets\traitlets.py", line 1348, in hold_trait_notifications
    value = trait._cross_validate(self, getattr(self, name))
  File "<env dir>\lib\site-packages\traitlets\traitlets.py", line 729, in _cross_validate
    value = obj._trait_validators[self.name](obj, proposal)
  File "<env dir>\lib\site-packages\traitlets\traitlets.py", line 1132, in __call__
    return self.func(*args, **kwargs)
  File "<dev dir>\jupyter_server\jupyter_server\services\contents\filemanager.py", line 79, in _validate_preferred_dir
    return super()._validate_preferred_dir(proposal)
  File "<dev dir>\jupyter_server\jupyter_server\services\contents\manager.py", line 89, in _validate_preferred_dir
    if not self.dir_exists(value):
  File "<dev dir>\jupyter_server\jupyter_server\services\contents\filemanager.py", line 187, in dir_exists
    os_path = self._get_os_path(path=path)
  File "<dev dir>\jupyter_server\jupyter_server\services\contents\fileio.py", line 262, in _get_os_path
    raise HTTPError(404, "%s is outside root contents directory" % path)
tornado.web.HTTPError: HTTP 404: Not Found (../example is outside root contents directory)
```

Reply:
I completely agree about improved error handling and didn't realize you were massaging an `HTTPError` into a `TraitError` in the overridden method.  This must be coming from the `self.dir_exists()` call in the super() method, so doing something like this in the super() would be sufficient to remove the override...
```python
        try:
            if not self.dir_exists(value):
                raise TraitError(_i18n("Preferred directory not found: %r") % value)
        except HTTPError as e:
            raise TraitError(e.log_message) from e
        return value
```
I just found it odd _only_ calling the superclass from an override.  Does `ContentsManager` have a contract that it doesn't raise `HTTPError`?  If so, then handling `HTTPError` only in the override makes more sense.

I'm fine leaving this as-is, although a comment might help future maintainers.

Reply:
I see your point. My original thought was that raising `HTTPError`s was a decision unique to `FileContentsManager` but it is probably a fair assumption that other file manager implementations might have copied this pattern as well, and that this therefore should be a general behavior. I will change it.